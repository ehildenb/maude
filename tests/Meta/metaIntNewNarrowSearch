#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/metaIntNewNarrowSearch.maude -no-banner -no-advise \
  > metaIntNewNarrowSearch.out 2>&1

diff $srcdir/metaIntNewNarrowSearch.expected metaIntNewNarrowSearch.out > /dev/null 2>&1
