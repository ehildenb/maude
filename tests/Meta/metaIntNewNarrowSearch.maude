set show timing off .
set show advisories off .

***
*** Test getNarrowingSearchResult()/getNarrowingSearchResultAndPath() in the meta-interpreter.
***

load metaInterpreter

mod NARROW is
  sort Foo .
  ops f g h : Foo -> Foo .
  ops i j k : Foo Foo -> Foo .

vars X Y Z W A B C D : Foo .
  eq j(f(X), Y) = i(f(Y), X) [variant] .
  rl g(i(X, k(Y, Z))) => f(k(Z, X)) [narrowing] .
endm

mod NARROW-TEST is
  pr META-INTERPRETER .
  sort MsgList .
  subsort Msg < MsgList .
  op empty : -> MsgList .
  op _;_ : MsgList MsgList -> MsgList [assoc id: empty] .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : MsgList -> Attribute .
  op start:_ : Term -> Attribute .
  op goal:_ : Term -> Attribute .
  op fold:_ : Qid -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  vars S S' : Substitution .
  var ST T G : Term .
  var TY : Type .
  var C : Context .
  var Q Q' L : Qid .
  var ML : MsgList .
  var TL : TermList .
  var F : Qid .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('NARROW, true)) .

  rl < X : User | AS, start: ST, goal: G, fold: F > insertedModule(X, Y) =>
     < X : User | AS, soln: 1, start: ST, goal: G, fold: F, got: empty >
     getNarrowingSearchResult(Y, X, 'NARROW, ST, G, '*, unbounded, F, 0) .

  rl < X : User | AS, soln: N, start: ST, goal: G, fold: F, got: ML >
     gotNarrowingSearchResult(X, Y, M, T, TY, S, Q, S', Q') =>
     < X : User | AS, soln: (N + 1),  start: ST, goal: G, fold: F,
     got: (ML ; gotNarrowingSearchResult(X, Y, M, T, TY, S, Q, S', Q')) >
     getNarrowingSearchResult(Y, X, 'NARROW, ST, G, '*, unbounded, F, N) .
endm

set show breakdown on .

erew in NARROW-TEST : <> 
< me : User | start: ('g['j['A:Foo, 'B:Foo]]),
goal: 'C:Foo,
fold: 'none >
createInterpreter(interpreterManager, me, none) .

erew in NARROW-TEST : <> 
< me : User | start: ('g['j['A:Foo, 'B:Foo]]),
goal: 'C:Foo,
fold: 'match >
createInterpreter(interpreterManager, me, none) .

mod NARROW-TEST is
  pr META-INTERPRETER .
  sort MsgList .
  subsort Msg < MsgList .
  op empty : -> MsgList .
  op _;_ : MsgList MsgList -> MsgList [assoc id: empty] .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : MsgList -> Attribute .
  op start:_ : Term -> Attribute .
  op goal:_ : Term -> Attribute .
  op fold:_ : Qid -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  vars S S' : Substitution .
  var ST T G : Term .
  var TY : Type .
  var C : Context .
  var Q Q' L : Qid .
  var ML : MsgList .
  var TL : TermList .
  var F : Qid .
  var TR : NarrowingTrace .
  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('NARROW, true)) .

  rl < X : User | AS, start: ST, goal: G, fold: F > insertedModule(X, Y) =>
     < X : User | AS, soln: 1, start: ST, goal: G, fold: F, got: empty >
     getNarrowingSearchResultAndPath(Y, X, 'NARROW, ST, G, '*, unbounded, F, 0) .

  rl < X : User | AS, soln: N, start: ST, goal: G, fold: F, got: ML >
     gotNarrowingSearchResultAndPath(X, Y, M, T, TY, S, TR, S', Q') =>
     < X : User | AS, soln: (N + 1),  start: ST, goal: G, fold: F,
     got: (ML ; gotNarrowingSearchResultAndPath(X, Y, M, T, TY, S, TR, S', Q')) >
     getNarrowingSearchResultAndPath(Y, X, 'NARROW, ST, G, '*, unbounded, F, N) .
endm

erew in NARROW-TEST : <> 
< me : User | start: ('g['j['A:Foo, 'B:Foo]]),
goal: 'C:Foo,
fold: 'none >
createInterpreter(interpreterManager, me, none) .

erew in NARROW-TEST : <> 
< me : User | start: ('g['j['A:Foo, 'B:Foo]]),
goal: 'C:Foo,
fold: 'match >
createInterpreter(interpreterManager, me, none) .
