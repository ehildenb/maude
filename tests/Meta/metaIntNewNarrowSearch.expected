==========================================
erewrite in NARROW-TEST : <> < me : User | start: ('g['j['A:Foo,'B:Foo]]),goal:
    'C:Foo,fold: 'none > createInterpreter(interpreterManager, me, none) .
rewrites: 13
mb applications: 0  equational rewrites: 4  rule rewrites: 5  variant narrowing
    steps: 3  narrowing steps: 1
result Configuration: <> < me : User | soln: 4,got: (gotNarrowingSearchResult(
    me, interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, '#, 
  '#1:Foo <- '@1:Foo ; 
  '#2:Foo <- '@2:Foo ; 
  'C:Foo <- 'g['j['@1:Foo,'@2:Foo]], '@) ; gotNarrowingSearchResult(me,
    interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, '#, 
  '#1:Foo <- 'f['%2:Foo] ; 
  '#2:Foo <- '%1:Foo ; 
  'C:Foo <- 'g['i['f['%1:Foo],'%2:Foo]], '%) ; gotNarrowingSearchResult(me,
    interpreter(0), 3, 'f['k['%2:Foo,'f['%3:Foo]]], 'Foo, 
  'A:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
  'B:Foo <- '%3:Foo, '%, 
  '%2:Foo <- '@1:Foo ; 
  '%3:Foo <- '@2:Foo ; 
  'C:Foo <- 'f['k['@1:Foo,'f['@2:Foo]]], '@)),start: ('g['j['A:Foo,'B:Foo]]),
    goal: 'C:Foo,fold: 'none > noSuchResult(me, interpreter(0), 4, true)
==========================================
erewrite in NARROW-TEST : <> < me : User | start: ('g['j['A:Foo,'B:Foo]]),goal:
    'C:Foo,fold: 'match > createInterpreter(interpreterManager, me, none) .
rewrites: 13
mb applications: 0  equational rewrites: 4  rule rewrites: 5  variant narrowing
    steps: 3  narrowing steps: 1
result Configuration: <> < me : User | soln: 4,got: (gotNarrowingSearchResult(
    me, interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, '#, 
  '#1:Foo <- '@1:Foo ; 
  '#2:Foo <- '@2:Foo ; 
  'C:Foo <- 'g['j['@1:Foo,'@2:Foo]], '@) ; gotNarrowingSearchResult(me,
    interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, '#, 
  '#1:Foo <- 'f['%2:Foo] ; 
  '#2:Foo <- '%1:Foo ; 
  'C:Foo <- 'g['i['f['%1:Foo],'%2:Foo]], '%) ; gotNarrowingSearchResult(me,
    interpreter(0), 3, 'f['k['%2:Foo,'f['%3:Foo]]], 'Foo, 
  'A:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
  'B:Foo <- '%3:Foo, '%, 
  '%2:Foo <- '@1:Foo ; 
  '%3:Foo <- '@2:Foo ; 
  'C:Foo <- 'f['k['@1:Foo,'f['@2:Foo]]], '@)),start: ('g['j['A:Foo,'B:Foo]]),
    goal: 'C:Foo,fold: 'match > noSuchResult(me, interpreter(0), 4, true)
==========================================
erewrite in NARROW-TEST : <> < me : User | start: ('g['j['A:Foo,'B:Foo]]),goal:
    'C:Foo,fold: 'none > createInterpreter(interpreterManager, me, none) .
rewrites: 13
mb applications: 0  equational rewrites: 4  rule rewrites: 5  variant narrowing
    steps: 3  narrowing steps: 1
result Configuration: <> < me : User | soln: 4,got: (
    gotNarrowingSearchResultAndPath(me, interpreter(0), 1, 'g['j['#1:Foo,
    '#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, nil, 
  '#1:Foo <- '@1:Foo ; 
  '#2:Foo <- '@2:Foo ; 
  'C:Foo <- 'g['j['@1:Foo,'@2:Foo]], '@) ; gotNarrowingSearchResultAndPath(me,
    interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, nil, 
  '#1:Foo <- 'f['%2:Foo] ; 
  '#2:Foo <- '%1:Foo ; 
  'C:Foo <- 'g['i['f['%1:Foo],'%2:Foo]], '%) ; gotNarrowingSearchResultAndPath(
    me, interpreter(0), 3, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, 
{
  [],
  ',
    '#1:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
    '#2:Foo <- '%3:Foo ; 
    'X:Foo <- 'f['%3:Foo] ; 
    'Y:Foo <- '%1:Foo ; 
    'Z:Foo <- '%2:Foo,
  '%,
  'f['k['%2:Foo,'f['%3:Foo]]],'Foo,
    'A:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
    'B:Foo <- '%3:Foo
}, 
  '%2:Foo <- '@1:Foo ; 
  '%3:Foo <- '@2:Foo ; 
  'C:Foo <- 'f['k['@1:Foo,'f['@2:Foo]]], '@)),start: ('g['j['A:Foo,'B:Foo]]),
    goal: 'C:Foo,fold: 'none > noSuchResult(me, interpreter(0), 4, true)
==========================================
erewrite in NARROW-TEST : <> < me : User | start: ('g['j['A:Foo,'B:Foo]]),goal:
    'C:Foo,fold: 'match > createInterpreter(interpreterManager, me, none) .
rewrites: 13
mb applications: 0  equational rewrites: 4  rule rewrites: 5  variant narrowing
    steps: 3  narrowing steps: 1
result Configuration: <> < me : User | soln: 4,got: (
    gotNarrowingSearchResultAndPath(me, interpreter(0), 1, 'g['j['#1:Foo,
    '#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, nil, 
  '#1:Foo <- '@1:Foo ; 
  '#2:Foo <- '@2:Foo ; 
  'C:Foo <- 'g['j['@1:Foo,'@2:Foo]], '@) ; gotNarrowingSearchResultAndPath(me,
    interpreter(0), 1, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, nil, 
  '#1:Foo <- 'f['%2:Foo] ; 
  '#2:Foo <- '%1:Foo ; 
  'C:Foo <- 'g['i['f['%1:Foo],'%2:Foo]], '%) ; gotNarrowingSearchResultAndPath(
    me, interpreter(0), 3, 'g['j['#1:Foo,'#2:Foo]], 'Foo, 
  'A:Foo <- '#1:Foo ; 
  'B:Foo <- '#2:Foo, 
{
  [],
  ',
    '#1:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
    '#2:Foo <- '%3:Foo ; 
    'X:Foo <- 'f['%3:Foo] ; 
    'Y:Foo <- '%1:Foo ; 
    'Z:Foo <- '%2:Foo,
  '%,
  'f['k['%2:Foo,'f['%3:Foo]]],'Foo,
    'A:Foo <- 'f['k['%1:Foo,'%2:Foo]] ; 
    'B:Foo <- '%3:Foo
}, 
  '%2:Foo <- '@1:Foo ; 
  '%3:Foo <- '@2:Foo ; 
  'C:Foo <- 'f['k['@1:Foo,'f['@2:Foo]]], '@)),start: ('g['j['A:Foo,'B:Foo]]),
    goal: 'C:Foo,fold: 'match > noSuchResult(me, interpreter(0), 4, true)
Bye.
